### PR TITLE
fix(verl): normalize per-rollout mean loss by rollout count

### LINKED ISSUE #593: per_rollout_mean divides by training rows instead of rollouts, so the loss scale changes when a rollout spans several rows
`per_rollout_mean` is the default loss mode (`agentlightning/verl/config.yaml:39`). In `trainer.py:617-626`
the advantages are rescaled once for the whole batch before the actor update:

```python
normalize_advantages_by_rollout(
    batch.batch["advantages"],
    batch.batch["response_mask"],
    rollout_ids,
    num_trained_rows=len(batch),   # rows, not rollouts
)
```

`normalize_advantages_by_rollout` (`per_rollout_loss.py:16-41`) then divides each row by
`rollout_token_count * num_trained_rows`.

That is exactly the per-rollout mean when a rollout occupies a single row. But a rollout can span
several rows: `trace_aggregator.level: transition` emits one row per turn, and even the default
`trajectory` level flushes an extra row when a turn's prompt does not extend the running context
(`rollout_adapter.py:542-571`, reported as `training/n_trace_merge_mismatch_rows`). In that case the
divisor counts rows, so every rollout keeps `1 / n_rows` instead of `1 / n_rollouts` and the batch
total becomes `n_rollouts / n_rows`.

## Reproduction

Calling the real adapter and then the same normalization `trainer.py` performs, on `main` @ `218f1f7`:

```text
[A] 2 rollouts, 1 row each   -> per-rollout mass {r1: 0.5,    r2: 0.5},   total = 1.0000
[B] 2 rollouts, 3 rows total -> per-rollout mass {r1: 0.3333, r2: 0.3333}, total = 0.6667
    (r1 hit a trace-merge mismatch; the adapter reports n_trace_merge_mismatch_rows = 1)
```

Case B differs from A only by a trace-merge mismatch, yet the whole batch carries 2/3 of the signal.
The same effect appears in isolation:

```python
normalize_advantages_by_rollout(ones(4, 3), ones(4, 3, dtype=long), ["A", "A", "A", "B"],
                                num_trained_rows=4)
# A mass 0.25, B mass 0.25, total 0.5
```

## Question

Is the divisor the number of rollouts or the number of rows? The current test
`tests/verl/test_per_rollout_loss.py::test_normalize_advantages_by_rollout` pins the row-based value
(2 rollouts over 3 rows keep `1/3` each), so I did not want to assume it is a bug.

* If it is meant to be the per-rollout mean: deriving the divisor from the distinct `rollout_ids`
  makes a rollout keep `1 / n_rollouts` regardless of how it was split. I have that patch ready
  (`per_rollout_loss.py`, `trainer.py`, and the test expectation `1/3` → `1/2`, plus a regression test
  over 1/2/3 rows per rollout). Full suite passes: 101 passed.
* If row-based scaling is intended: the docstring ("`rollout's token count and batch size`") and the
  test would be worth stating explicitly, since switching `trace_aggregator.level` then silently
  rescales the loss.

Happy to open the PR for the first option, or rework it, once you confirm the intended semantics.

### PR DESCRIPTION
**Pull request by KunyangZhang**
## What

`normalize_advantages_by_rollout` divides each training row by
`rollout_token_count * num_trained_rows`, and `trainer.py` passes
`num_trained_rows=len(batch)` — the number of training **rows**. One rollout can occupy several
rows (`trace_aggregator.level: transition`, or a trace-merge mismatch at the default `trajectory`
level), so a rollout with `k` rows keeps `1 / n_rows` instead of `1 / n_rollouts` and the batch
total becomes `n_rollouts / n_rows`.

This PR derives the divisor from the distinct `rollout_ids`, so each rollout keeps exactly
`1 / n_rollouts` of the advantage mass no matter how it was split.

## Why

The v1.0 technical report (arXiv:2608.17528, §2.3, linked from
`docs/20-trainer-configuration.md`) defines this loss as the rollout-level token-mean, Eq. 16:

```
L_rollout-mean = (1/R) * sum_rho [ (sum_j sum_t l_rho,j,t) / (sum_j L_rho,j) ]
```

and states that *"sample count should not be allowed to affect gradient normalization, because it
is often driven by incidental factors such as retokenization"*. The report's own SWE-bench runs
produce 2.41 samples per rollout on average, with only 36% of rollouts staying a single merged row.

On the report's Figure 5 batch (A: 50 + 100 tokens, B: 30 x 3, C: 40) with `l = 1` per token, the
current code carries a batch total of `0.5` where Eq. 16 gives `1.0`; with the 2.41 samples/rollout
measured in the paper the factor is ≈ `1/2.41 ≈ 0.41`, and it moves with samples-per-rollout.

Fixes #593.

## How verified

```bash
uv sync --frozen --no-default-groups --extra dev --group dev --group verl-cpu
.venv/bin/python -m pytest -q tests        # 101 passed (100 before this change)
ruff check / ruff format --check / scripts/check_headers.py   # clean
```

Two things worth reviewing:

* `tests/verl/test_per_rollout_loss.py::test_normalize_advantages_by_rollout` pinned the old
  expectation (2 rollouts over 3 rows keeping `1/3` each); it now expects `1/2`, and a new
  `test_normalize_advantages_by_rollout_is_row_count_invariant` walks 1/2/3 rows per rollout and
  asserts the `n_rollouts`-invariance of the per-rollout mass.
* The updated signature drops the `num_trained_rows` keyword. If you would rather keep it for
  compatibility and ignore it, I am happy to adjust.

Prepared with AI assistance; the change was reviewed and verified locally by the author.

### CODE REVIEW

**Comment by KunyangZhang** [comment]
Withdrawing this. Siwei explained on #593 that dividing by the collected row count `N` rather than
the distinct rollout count is deliberate: it compensates the `N/R` extra mini-batch update units,
so the report-level `1/R` and this `R/N` factor compose to `1/N` and cancel at the training-update
level. I confirmed the second factor in the code (`trainer.py:502-507`: `n_remained_transition`
derives from `len(batch)`, and `max_ppo_update_times` caps exactly that count), so this patch
removes an intentional correction and should not be merged.

Thanks for the detailed explanation — I have followed up on #593 with the one documentation gap I
noticed while working through it.
